src/main/python/pgshovel/replication/streams/kafka.py

```diff
@@ -39,15 +55,26 @@ def consume(self, state):
             logger.debug('Moving to previous replication log offset: %s (current position: %s)...', offset, current)
             consumer.seek(delta, 1)
             assert consumer.offsets[0] == offset
+        else:
+            decoded = prime_for_batch_start(
+                decoded,
+                max_messages=self.prime_threshold
+            )
 
-        for offset, message in consumer:
-            decoded = self.codec.decode(message.value)
-            state = validate_state(state, offset, decoded)
+        for offset, message in decoded:
+            state = validate_state(state, offset, message)
             # XXX: This is necessary because of a bug in protocol buffer oneof.
             state = type(state).FromString(state.SerializeToString())
-            yield state, offset, decoded
+            yield state, offset, message
+
 
     @classmethod
     def configure(cls, configuration, cluster, set):
         topic = '{cluster}.{set}.mutations'.format(cluster=cluster.name, set=set)
-        return cls(cluster, set, configuration['hosts'], topic)
+        return cls(
+            cluster,
+            set,
+            configuration['hosts'],
+            topic,
+            configuration.get('prime_threshold', 100)
```

Comment:
This actually might be a bit low for a default, we were routinely seeing batches containing 200-400 individual operations per batch. Since a very large operation (`DELETE FROM table`) could affect a large number of rows, I almost wonder if priming should be allowed to be unbounded. I'm not sure if that's a sane default or not, but since the batch size is going to potentially vary heavily from application to application, it seems like it might be better to not be too prescriptive here?